### PR TITLE
Add swap support

### PR DESCRIPTION
**Pull request by teawater**
Add swap support can help to increase memory capacity, handle some memory issues and increase file access speed sometimes.
These patches add swap support to kata.
When option "enable_guest_swap" in hypervisor.qemu is enabled, kata will insert a raw file to the guest as the swap device if the swappiness of a container (set by annotation "io.katacontainers.resource.swappiness") is bigger than 0.
Use annotation "io.katacontainers.resource.swappiness" but not c.config.Resources.Memory.Swappiness because containerd doesn't support it.  The "io.katacontainers.resource.swap_in_bytes" is same.

I will make special PR for doc and test when this PR is merged.

Fixes: #2201

### CODE REVIEW

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -1192,6 +1196,16 @@ impl protocols::agent_ttrpc::AgentService for AgentService {
 
         Err(ttrpc_error(ttrpc::Code::INTERNAL, ""))
     }
+
+    async fn add_swap(
+        &self,
+        _ctx: &TtrpcContext,
```
Comment (liubin) [inline src/agent/src/rpc.rs:1202]: `_ctx` -> `ctx` and create a span with `trace_rpc_call!(ctx, "add_swap", req);` , like other RPC calls.

File: src/runtime/cli/config/configuration-qemu.toml.in [hunk 2]
```diff
@@ -356,6 +356,16 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
 # See: https://www.qemu.org/docs/master/qemu-qmp-ref.html#Dump-guest-memory for details
 #guest_memory_dump_paging=false
 
+# Enable swap in the guest. Default false.
+# When enable_guest_swap is enabled, insert a raw file to the guest as the swap device
+# if the swappiness of a container (set by annotation "io.katacontainers.resource.swappiness")
+# is bigger than 0.
+# The size of the swap device should be 
+# swap_in_bytes (set by annotation "io.katacontainers.resource.swap_in_bytes") - memory_limit_in_bytes.
```
Comment (liubin) [inline src/runtime/cli/config/configuration-qemu.toml.in:364]: Should these annotation keys be documented in `how-to-set-sandbox-config-kata.md` ?

File: src/runtime/cli/config/configuration-qemu.toml.in [hunk 3]
```diff
@@ -356,6 +356,16 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
...
+# If swap_in_bytes is not set, the size should be memory_limit_in_bytes.
+# If swap_in_bytes and swap_in_bytes is not set, the size should be default_memory.
```
Comment (liubin) [inline src/runtime/cli/config/configuration-qemu.toml.in:366]: `If swap_in_bytes and swap_in_bytes is not set`, there are double `swap_in_bytes`.

File: src/runtime/virtcontainers/pkg/oci/utils.go [hunk 4]
```diff
@@ -616,6 +617,15 @@ func addHypervisorMemoryOverrides(ocispec specs.Spec, sbConfig *vc.SandboxConfig
 		return err
 	}
 
+	if value, ok := ocispec.Annotations[vcAnnotations.EnableGuestSwap]; ok {
```
Comment (liubin) [inline src/runtime/virtcontainers/pkg/oci/utils.go:620]: Setting bool values can be changed to like lines 614-618.

File: src/runtime/virtcontainers/sandbox.go [hunk 5]
```diff
@@ -995,6 +1004,104 @@ func (cw *consoleWatcher) stop() {
 	}
 }
 
+func (s *Sandbox) addSwap(ctx context.Context, id string, size int64) (*config.BlockDrive, error) {
```
Comment (liubin) [inline src/runtime/virtcontainers/sandbox.go:1007]: NIT: `id` change something more details, like `swapID` or `deviceID`.

File: src/runtime/virtcontainers/sandbox.go [hunk 6]
```diff
@@ -995,6 +1004,104 @@ func (cw *consoleWatcher) stop() {
...
+	defer func() {
+		if err != nil {
+			os.Remove(swapFile)
+		}
+	}()
+
+	err = os.Truncate(swapFile, size)
+	if err != nil {
+		err = fmt.Errorf("truncate swapfile %s fail %s", swapFile, err.Error())
+		s.Logger().Error(err)
```
Comment (liubin) [inline src/runtime/virtcontainers/sandbox.go:1026]: NIT: Log should be `.WithError(err).Error("some info")`

**Comment by amshinde** [comment]
@teawater Can you explain what scenario you are trying to address here? I am not sure adding swap support will play well with k8s resource constraints and limits and may further mess up with resource accounting.
cc @egernst

File: src/runtime/cli/config/configuration-qemu.toml.in (hunk 2 above)
Comment (teawater) [reply src/runtime/cli/config/configuration-qemu.toml.in:364]: We just have two annotations "io.katacontainers.resource.swap_in_bytes" and "io.katacontainers.resource.swappiness" for container.
Both of them are the swap options.
Do you mind I introduce them in the doc for swap when I add the doc?

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (teawater) [reply src/agent/src/rpc.rs]: Added.

File: src/runtime/cli/config/configuration-qemu.toml.in (hunk 3 above)
Comment (teawater) [reply src/runtime/cli/config/configuration-qemu.toml.in]: Fixed.

File: src/runtime/virtcontainers/pkg/oci/utils.go (hunk 4 above)
Comment (teawater) [reply src/runtime/virtcontainers/pkg/oci/utils.go]: Fixed.

File: src/runtime/virtcontainers/sandbox.go (hunk 5 above)
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go]: I changed it to swapID.

File: src/runtime/virtcontainers/sandbox.go (hunk 6 above)
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go:1026]: I set swapFile and err.Error() to err than they can return back to caller with err.
Do you mind I keep this part?

**Comment by teawater** [comment]
> @teawater Can you explain what scenario you are trying to address here? I am not sure adding swap support will play well with k8s resource constraints and limits and may further mess up with resource accounting.
> cc @egernst

Add swap support can help to increase memory capacity, handle some memory issues and increase file access speed sometimes.
I am not sure about k8s resource because I'm not very familiar with it.
But container has long history with swap support.  That is why  config of swap available in https://github.com/kata-containers/kata-containers/blob/main/src/runtime/vendor/github.com/opencontainers/runtime-spec/specs-go/config.go#L296

**Comment by fidencio** [comment]
Hey, we've had yet another breakage in our CI and now it's fixed as part of https://github.com/kata-containers/kata-containers/pull/2203 ... it means that you'll need to rebase your PR atop of current main.

Sorry for the inconvenience.

File: src/agent/src/rpc.rs [hunk 7]
```diff
@@ -1541,6 +1557,56 @@ fn do_copy_file(req: &CopyFileRequest) -> Result<()> {
...
+        root_bus_sysfs,
+        pcipath_to_sysfs(&root_bus_sysfs, &pcipath)?
+    );
+    let (mut virtio_path, _) = path_name_lookup(sysfs_rel_path, "virtio")?;
+    virtio_path.push("block");
+    let (_, dev_name) = path_name_lookup(virtio_path, "vd")?;
+    let dev_name = format!("/dev/{}", dev_name);
+
+    let c_str = CString::new(dev_name)?;
+    let ret = unsafe { libc::swapon(c_str.as_ptr() as *const c_char, 0) };
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1599]: There isn't a corresponding call to `swapoff()`, but the swap device is deleted, which might cause bad results on shutdown?

File: src/runtime/virtcontainers/sandbox.go [hunk 8]
```diff
@@ -1816,9 +1933,21 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
...
-	sandboxMemoryByte += int64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
+	hypervisorMemoryByte := int64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
+	sandboxMemoryByte += hypervisorMemoryByte
+	if sandboxneedPodSwap {
+		sandboxSwapByte += hypervisorMemoryByte
+	}
+
+	// Setup the SWAP in the guest
+	if sandboxSwapByte > 0 {
+		err = s.setupSwap(ctx, sandboxSwapByte)
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/sandbox.go:1946]: According to `mkswap(8)`,

>  The maximum useful size of a swap area depends on the architecture and the kernel version.

Shame it doesn't actually tell us _what_ the limits are though ;( Realistically, I suspect we'll never hit them but just an fyi.

File: src/runtime/virtcontainers/sandbox.go [hunk 9]
```diff
@@ -1859,23 +1988,45 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
 	if err := s.agent.onlineCPUMem(ctx, 0, false); err != nil {
 		return err
 	}
+
 	return nil
 }
 
-func (s *Sandbox) calculateSandboxMemory() int64 {
+func (s *Sandbox) calculateSandboxMemory() (int64, bool, int64) {
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/sandbox.go:1995]: According to `mkswap(8)`,

>  mkswap refuses areas smaller than 10 pages.

Is this handled by the function below? Might be worth adding more unit tests to ensure the minimum swap size is atleast 1MiB or similar? In fact, it might be worth moving the new logic in this function into a new function to make this simpler?

File: src/runtime/virtcontainers/container.go [hunk 10]
```diff
@@ -43,6 +44,11 @@ func (c *Container) tracingTags() map[string]string {
 	}
 }
 
+const (
+	resourceSwappinessLabel  = "io.katacontainers.resource.swappiness"
+	resourceSwapInBytesLabel = "io.katacontainers.resource.swap_in_bytes"
+)
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/container.go:50]: Shouldn't these new annotations be defined in `src/runtime/virtcontainers/pkg/annotations/annotations.go`?

File: src/runtime/virtcontainers/container.go [hunk 11]
```diff
@@ -767,6 +773,25 @@ func newContainer(ctx context.Context, sandbox *Sandbox, contConfig *ContainerCo
...
+		if err == nil && resourceSwappiness > 200 {
+			err = fmt.Errorf("swapiness should not bigger than 200")
+		}
+		if err != nil {
+			return &Container{}, fmt.Errorf("Invalid container configuration Annotations %s %v", resourceSwappinessLabel, err)
+		}
+		c.config.Resources.Memory.Swappiness = &resourceSwappiness
+	}
+	if resourceSwapInBytesStr, ok := c.config.Annotations[resourceSwapInBytesLabel]; ok {
+		resourceSwapInBytes, err := strconv.ParseInt(resourceSwapInBytesStr, 0, 64)
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/container.go:788]: This looks unsafe - what if `io.katacontainers.resource.swap_in_bytes = "-1"`? `ParseUint()`?

**Review by jodh-intel — COMMENTED** [review]
Thanks @teawater - a few initial comments...

File: src/agent/src/rpc.rs (hunk 7 above)
Comment (teawater) [reply src/agent/src/rpc.rs:1599]: I didn't do swapoff because all the data inside the swap device will go back to memory when swapoff.
It will crash the application in the guest kernel because it will use a lot of memory in the guest system.
So I think the better way to stop swap is shutdown the vm and remove the swap file.

File: src/runtime/cli/config/configuration-qemu.toml.in (hunk 2 above)
Comment (teawater) [reply src/runtime/cli/config/configuration-qemu.toml.in]: I added a special commit that add swap annotations introduction of
    `io.katacontainers.config.hypervisor.enable_guest_swap`,
    `io.katacontainers.container.resource.swappiness"` and
    `io.katacontainers.container.resource.swap_in_bytes"` to
    how-to-set-sandbox-config-kata.md.

File: src/runtime/virtcontainers/sandbox.go (hunk 6 above)
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go]: Done.

File: src/runtime/virtcontainers/sandbox.go [hunk 12]
```diff
@@ -992,6 +1001,104 @@ func (cw *consoleWatcher) stop() {
...
+	}()
+
+	err = os.Truncate(swapFile, size)
+	if err != nil {
+		err = fmt.Errorf("truncate swapfile %s fail %s", swapFile, err.Error())
+		s.Logger().Error(err)
+		return nil, err
+	}
+
+	err = exec.CommandContext(ctx, "mkswap", swapFile).Run()
```
[the remark this answers is not in the document]
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go]: Fixed.

File: src/runtime/virtcontainers/sandbox.go (hunk 9 above)
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go:2002]: I updated function addSwap to handle the size issue.
```
	// Check the size
	pagesize := os.Getpagesize()
	// mkswap refuses areas smaller than 10 pages.
	size = int64(math.Max(float64(size), float64(pagesize*10)))
	// Swapfile need a page to store the metadata
	size += int64(pagesize)
```

File: src/runtime/virtcontainers/container.go (hunk 10 above)
Comment (teawater) [reply src/runtime/virtcontainers/container.go]: Moved to annotations.go and rename them to io.katacontainers.container.resource.swappiness and io.katacontainers.container.resource.swap_in_bytes.

File: src/runtime/virtcontainers/container.go (hunk 11 above)
Comment (teawater) [reply src/runtime/virtcontainers/container.go]: Fixed.

File: docs/how-to/how-to-set-sandbox-config-kata.md [hunk 13]
```diff
@@ -91,6 +91,13 @@ There are several kinds of Kata configurations and they are listed below.
 | `io.katacontainers.config.hypervisor.virtio_fs_cache` | string | the cache mode for virtio-fs, valid values are `always`, `auto` and `none` |
 | `io.katacontainers.config.hypervisor.virtio_fs_daemon` | string | virtio-fs `vhost-user` daemon path |
 | `io.katacontainers.config.hypervisor.virtio_fs_extra_args` | string | extra options passed to `virtiofs` daemon |
+| `io.katacontainers.config.hypervisor.enable_guest_swap` | `boolean` | enable swap in the guest |
+
+## Container Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.container.resource.swappiness"` | `uint64` | specify the Resources.Memory.Swappiness |
+| `io.katacontainers.container.resource.swap_in_bytes"` | `int64` | specify the Resources.Memory.Swap |
```
Comment (jodh-intel) [inline docs/how-to/how-to-set-sandbox-config-kata.md:100]: `uint64`.

File: src/runtime/virtcontainers/sandbox.go [hunk 14]
```diff
@@ -992,6 +1001,111 @@ func (cw *consoleWatcher) stop() {
...
+	size += int64(pagesize)
+
+	err = os.Truncate(swapFile, size)
+	if err != nil {
+		err = fmt.Errorf("truncate swapfile %s fail %s", swapFile, err.Error())
+		s.Logger().WithError(err).Error("addSwap")
+		return nil, err
+	}
+
+	err = exec.CommandContext(ctx, "/sbin/mkswap", swapFile).Run()
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/sandbox.go:1034]: - Might be worth making `/sbin/mkswap` a top-level `const` for maximum visibility.
- Unlike `swapon(8)`, if this fails, there are no useful return values (it's just `0` or `1`). As such, I think we should capture stdout+stderr and log these on error. Something like:
  ```go
  var outbuf, errbuf bytes.Buffer
  cmd := exec.CommandContext(ctx, "/sbin/mkswap", swapFile)
  cmd.Stdout = &outbuf
  cmd.Stderr = &outbuf
  err := cmd.Run()
  ```
  In fact, we should call all external commands using a wrapper function that does this imho.

**Review by jodh-intel — COMMENTED** [review]
Thanks @teawater.

A few more comments.

**Comment by fidencio** [comment]
I'm adding the `do-not-merge` label till we get `2.2.0-alpha1` release out.
Once we cut the release I'll come back removing the label from the PRs.

File: docs/how-to/how-to-set-sandbox-config-kata.md (hunk 13 above)
Comment (teawater) [reply docs/how-to/how-to-set-sandbox-config-kata.md]: Fixed.

File: src/runtime/virtcontainers/sandbox.go (hunk 14 above)
Comment (teawater) [reply src/runtime/virtcontainers/sandbox.go]: Added.

**Comment by c3d** [comment]
I see annotations documented.
Is there also a confit file option
(Watching from my phone on the road, apologies if I missed it)

**Comment by amshinde** [comment]
> Add swap support can help to increase memory capacity, handle some memory issues and increase file access speed sometimes.
> I am not sure about k8s resource because I'm not very familiar with it.
> But container has long history with swap support. That is why config of swap available in https://github.com/kata-containers/kata-containers/blob/main/src/runtime/vendor/github.com/opencontainers/runtime-spec/specs-go/config.go#L296

@teawater Afaik with k8s you need to disable swap on the host system, part of this is for memory accounting. I see that the annotations exist for swappiness in the runtime spec, but how is this feature supported in k8s/containerd?

**Comment by teawater** [comment]
> I see annotations documented.
> Is there also a confit file option
> (Watching from my phone on the road, apologies if I missed it)

I will make special PR for doc and test when this PR is merged.

**Comment by teawater** [comment]
> > Add swap support can help to increase memory capacity, handle some memory issues and increase file access speed sometimes.
> > I am not sure about k8s resource because I'm not very familiar with it.
> > But container has long history with swap support. That is why config of swap available in https://github.com/kata-containers/kata-containers/blob/main/src/runtime/vendor/github.com/opencontainers/runtime-spec/specs-go/config.go#L296
> 
> @teawater Afaik with k8s you need to disable swap on the host system, part of this is for memory accounting. I see that the annotations exist for swappiness in the runtime spec, but how is this feature supported in k8s/containerd?

This PR will enable swap in the guest.

**Comment by fidencio** [comment]
I'm really sorry for the disruption we had, but I'm glad to say that 2.2.0-alpha1 release is finally out and we're removing all the `do-not-merge` labels that were added due to this.

Thanks a lot for your understanding and happy hacking!

**Review by liubin — APPROVED** [review]

**Comment by teawater** [comment]
Got a issue with the doc commit that about markdown because check-markdown doesn't support word "Swappiness".
Remove the doc commit from the current PR.

I will make special PR for doc(should add support of "Swappiness" to check-markdown first ) and test when this PR is merged.

**Comment by teawater** [comment]
jenkins-ci-ARM-ubuntu-18-04 got:
```
time="2021-07-20T08:57:51.214366559Z" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:sandbox,Uid:81855ad8681d0d86d1e91e00167939cb6694d2c422acd208a0072939487f6999,Namespace:image-load-52fdfc072182654f163f5f0f9a621d729566c74d10037c4d7bbb0407d1e2c649-eb9d18a44784045d87f3c67cf22746e995af5a25367951baa2ff6cd471c483f1,Attempt:0,} failed, error" error="failed to setup network for sandbox \"db1e54d96580e034aafcb08a1fda7cc03208aa6b5aaf1b8fe465ee318950e917\": failed to set bridge addr: \"cni0\" already has an IP address different from 10.88.0.1/16"
time="2021-07-20T08:57:51.285052393Z" level=debug msg="received signal" signal=terminated
time="2021-07-20T08:57:51.285458088Z" level=debug msg="sd notification" error="<nil>" notified=false state="STOPPING=1"
time="2021-07-20T08:57:51.285900034Z" level=info msg="Stop CRI service"
time="2021-07-20T08:57:51.287119520Z" level=debug msg="cni watcher channel is closed"
time="2021-07-20T08:57:51.287152545Z" level=info msg="Stop CRI service"
```
This failure is not related to the current PR.

**Review by Tim-Zhang — APPROVED** [review]

**Comment by bergwolf** [comment]
Thanks @teawater !
